### ISSUE TITLE
i686-w64-mingw32-gcc assumes 16 byte stack alignment

### ISSUE THREAD

**Issue by stefand** [mingw-w64/mingw-w64#30]
Mingw targeting x86 Windows incorrectly assumes the incoming stack alignment is 16 byte. To reproduce, build the attached test.txt (it is a C source, just rename) and disassemble it:
[test.txt](https://github.com/mingw-w64/mingw-w64/files/12387975/test.txt)

```
$ i686-w64-mingw32-gcc test.c -c -O2
$ i686-w64-mingw32-objdump -d test.o 

   0:   81 ec ac 01 00 00       sub    $0x1ac,%esp
   6:   8d 44 24 10             lea    0x10(%esp),%eax
   a:   89 04 24                mov    %eax,(%esp)
   d:   e8 00 00 00 00          call   12 <_func+0x12>
  12:   81 c4 ac 01 00 00       add    $0x1ac,%esp
  18:   c3                      ret

```

The structure requests 16 byte alignment, but this only works if the incoming stack is 16 byte aligned. This is not the case according to the Win32 calling convention.

x86_64-w64-mingw32-gcc when invoked with -m32 does the correct thing:

```
$ x86_64-w64-mingw32-gcc test.c -c -O2 -m32
$ i686-w64-mingw32-objdump -d test.o
   0:   55                      push   %ebp
   1:   89 e5                   mov    %esp,%ebp
   3:   83 e4 f0                and    $0xfffffff0,%esp
   6:   81 ec a0 01 00 00       sub    $0x1a0,%esp
   c:   8d 44 24 10             lea    0x10(%esp),%eax
  10:   89 04 24                mov    %eax,(%esp)
  13:   e8 00 00 00 00          call   18 <_func+0x18>
  18:   c9                      leave
  19:   c3                      ret

```

We came across this bug in Wine because we have users who try to build with -O3 and the resulting code chokes over unaligned SSE memory reads. I have filed https://bugs.winehq.org/show_bug.cgi?id=55470 to keep track of this issue on Wine's side.

**Comment by lhmouse** [comment]
This issue is known and won't be fixed.

Callbacks from 32-bit system libraries might get a misaligned ESP, so they should be marked with the `force_align_arg_pointer` attribute, like
````c
__attribute__((__force_align_arg_pointer__)) int baz(int *foo);
````

**Comment by stefand** [comment]
I was actively searching for a statement like this. Either I am totally blind or this should be documented more prominently.

There are plenty of callbacks in your average Win32 program, including unexpected places. main() and WNDPROCs are well known. COM methods are less obvious. APIs like ddraw have callback functions all over the place.

IMHO the ABI shouldn't be broken by default. If someone wants to manually handle this, let them set -mincoming-stack-boundary=4 and audit the code for callbacks.

Some third parties wondering about the same issue I came across in my search:
https://stackoverflow.com/questions/76482839/g-mingw-x86-32-assumes-16-byte-stack-alignment-in-imgui-code
https://github.com/libsdl-org/SDL/pull/7607
https://github.com/nothings/stb/issues/81

(Yes, those are all triggered by explicitly enabling SSE, but SSE isn't the only place where this matters. My example deliberately does not use SSE intrinsics)

**Comment by Alcaro** [comment]
> This issue is known and won't be fixed.

With all due respect, that makes no sense whatsoever.

On i386 Linux, the ABI demands 16-byte stack alignment, and expecting a attribute((my_caller_is_buggy)) near noncompliant code is fine.

But the i386 Windows ABI only requires 4-byte alignment, and anything that claims to be a i386 Windows compiler should default to following the i386 Windows ABI, not the Linux ABI. i386-mingw-gcc should default to -mpreferred-stack-boundary=2.

Just like MSVC does, and just like x86_64-mingw-gcc does with -m32. <https://godbolt.org/z/sdh9bxM8h>

**Comment by lhmouse** [comment]
> But the i386 Windows ABI only requires 4-byte alignment, and anything that claims to be a i386 Windows compiler should default to following the i386 Windows ABI, not the Linux ABI. i386-mingw-gcc should default to -mpreferred-stack-boundary=2.

If you think it is necessary, please file a bug report at https://gcc.gnu.org/bugzilla/.

**Comment by stefand** [comment]
I investigated why CrossOver, which builds its i386 DLLs with -mfpmath=sse -march=core2 doesn't have problems. It doesn't explicitly align the stack, but it turns out that -march=core2 causes mingw to align it.

-march=pentium3 / -march=athlon-4 are the oldest CPUs that cause this. P2 or athlon-tbird do no. (I don't exactly know the order of releases of AMD CPUs)

I'll consult with some other Wine devs and file a gcc bug. What's the separation of concerns here between mingw and gcc? I blindly assumed that defining the Windows ABI is a matter of mingw, not upstream gcc.

**Comment by cjacek** [comment]
Yeah, it's a long standing issue with GCC, I hit this problem with Firefox build 13 years ago: https://bugzilla.mozilla.org/show_bug.cgi?id=631252. Back then Kai didn't agree with me that it's a GCC bug, so I gave up and worked around it in Firefox. In my opinion it's definitely a GCC bug and it would be great to get it fixed there.

> What's the separation of concerns here between mingw and gcc?

mingw-w64 is just about platform/crt headers and libraries. The compiler is a separated thing, mingw-64 it's not even limited to a single compiler vendor anymore.

**Comment by zfigura** [comment]
> This issue is known and won't be fixed.

Er, what? Upstream gcc *does* assume 4-byte stack alignment for the i686-w64-mingw32 target, at least when it comes to SSE instructions. It's just that it's apparently broken for __attribute__((aligned)).

**Comment by lhmouse** [comment]
> Er, what? Upstream gcc _does_ assume 4-byte stack alignment for the i686-w64-mingw32 target, at least when it comes to SSE instructions. It's just that it's apparently broken for **attribute**((aligned)).

The native GCC for i686-w64-mingw32 assumes 16-byte alignment. If an `__m256` variable is stored on the stack, GCC realigns ESP as needed (tested with GCC 13.2). It wasn't fixed because Kai did not think it was a GCC issue: https://sourceforge.net/p/mingw-w64/mailman/message/35027678/

The native GCC for x86_64-w64-mingw32 does not realign RSP in case of AVX, but that's another bug.

**Comment by rofl0r** [comment]
> I'll consult with some other Wine devs and file a gcc bug.

@stefand: would you mind posting the bugzilla URL ? i'd like to find a patch or a solution that i can hardcode into the specs file so i686-mingw-gcc does the right thing.

**Comment by lhmouse** [comment]
I see that this is no longer the case since GCC 13. Would you please check that?

**Comment by lhmouse** [comment]
It's because MSYS2 changed their default arch to Pentium 4:
 
* https://github.com/msys2/MINGW-packages/commit/0190f0e75cb02c3f676229a16fceaf275de6046b#diff-cf683a120abf18dd9f774ea5df77137d01e19f359f7cb40c1fc152355cde7b80R181-R186
* https://gcc.gnu.org/bugzilla/show_bug.cgi?id=111107#c19

**Comment by lhmouse** [comment]
This should be fixed by https://gcc.gnu.org/cgit/gcc/commit/?id=ee7c0a5b70dc316477f45abc0f09dd2af9abe5cb.

It's not in any GCC release yet.
